pkg/loop/internal/pb/chain_reader.proto

```diff
@@ -9,23 +9,51 @@ import "google/protobuf/empty.proto";
 
 service ChainReader {
   rpc GetLatestValue (GetLatestValueRequest) returns (GetLatestValueReply) {}
+  rpc QueryKey(QueryKeyRequest) returns (QueryKeyReply) {}
   rpc Bind(BindRequest) returns (google.protobuf.Empty) {}
 }
 
 // GetLatestValueRequest has arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.GetLatestValue].
 message GetLatestValueRequest {
-  string contractName = 1;
+  string contract_name = 1;
   string method = 2;
   VersionedBytes params = 3;
 }
 
+// QueryKeyRequest has arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.QueryKey].
+message QueryKeyRequest {
+  string contract_name = 1;
+  QueryKeyFilter filter = 2;
+  LimitAndSort limit_and_sort = 3;
+}
+
+// BindRequest has arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.Bind].
+message BindRequest {
+  repeated BoundContract bindings = 1;
+}
+
 // GetLatestValueReply has return arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.GetLatestValue].
 message GetLatestValueReply {
   VersionedBytes retVal = 1;
 }
 
-message BindRequest {
-  repeated BoundContract bindings = 1;
+// QueryKeyReply has return arguments for [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.QueryKey].
+message QueryKeyReply {
+  repeated Sequence sequences = 1;
+}
+
+// Head is gRPC adapter for the Head struct [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.Head].
+message Head {
+  string identifier = 1;
+  bytes hash = 2;
+  uint64 timestamp = 3;
+}
+
+// Head is gRPC adapter for the Head struct [github.com/smartcontractkit/chainlink-common/pkg/types.ChainReader.Sequence].
+message Sequence {
+  string sequence_cursor = 1;
```

Comment:
out of curiosity why is the cursor a string? is it like a nextPage token?

Reply:
EVM cursor would be "blockHash-txHash-logIndex" it will be used both as nextPage token, but also so that you can say give me 100 events that happened before this one